### DISCUSSION TITLE
/ORM/getting-started.html

### DISCUSSION THREAD

**Discussion by giscus** [goravel/goravel#311] [Announcements]
# /ORM/getting-started.html

[[toc]] 简介 Goravel 提供了一套非常简单易用的数据库交互方式，开发者可以使用 facades.Orm() 进行操作。目前，Goravel 为以下四种数据库提供了官方支持： MySQL 5.7+; PostgreSQL 9.6+; SQLite 3.8.8+; SQL Server 2017+; 在开始之前，请在 .env 文件中配置数据...

https://www.goravel.dev/zh/ORM/getting-started.html

**Comment by pello1020** [comment]
请问如何可以根据表字段 生成 model struct 文件，pgsql

**Reply by hwbrzzl** [reply]
需要用到第三方工具，群里好像有人处理过，可以入群咨询下。

**Reply by jiqinwj** [reply]
很复杂的

**Comment by nanaaikinson** [comment]
I've gone through the docs for goravel's ORM and its amazing. It will be great if this is a standalone package to be used with other non goravel projects/frameworks. In the PHP world, laravel database can be used without laravel itself. Will there be something like that for goravel's ORM or other components of goravel?

**Reply by hwbrzzl** [reply]
Yes, we can have it in the future, welcome PR to implement it.

**Comment by luscio** [comment]
laravel里模型追加属性append，在goravel要怎么实现。刚从php转过来，不太懂。

**Reply by hwbrzzl** [reply]
Sorry, Goravel 暂时还没有这个功能，预期之后会增加。

**Comment by xx1196** [comment]
Hi, how i can send the paginated data? i send de param page in postman but this return alwayes de first page, can help me? thanks

**Reply by hwbrzzl** [reply]
Please show your code.

**Comment by xx1196** [comment]
i resolve with this way:
```go
func (r *GenderController) Index(ctx http.Context) http.Response {
	page, err := strconv.Atoi(ctx.Request().Query("page", "1"))
	var genders []models.Gender
	var total int64

	err = facades.Orm().Query().Paginate(page, 2, &genders, &total)
	if err != nil {
		return ctx.Response().Json(http.StatusInternalServerError, http.Json{
			"error": err.Error(),
		})
	}
	return ctx.Response().Success().Json(http.Json{
		"data": genders,
	})
}
```

**Comment by xx1196** [comment]
Hello, in Laravel I can modify the information using mutators, in Goravel can I also?

**Reply by hwbrzzl** [reply]
Goravel can not do that yet, there is an issue to achieve it: https://github.com/goravel/goravel/issues/117

**Comment by klphp** [comment]
var exists bool
facades.Orm().Query().Where("name", "tom").Exists(&exists) 
查询的是哪个表，或者说，查询users表怎么设置

**Reply by hwbrzzl** [reply]
Here: https://www.goravel.dev/zh/orm/getting-started.html#%E6%8C%87%E5%AE%9A%E8%A1%A8%E6%9F%A5%E8%AF%A2

**Comment by ruoxiao-zh** [comment]
how can i get *gorm.DB

**Reply by ruoxiao-zh** [reply]
facades.Orm().Query().(*gorm.QueryImpl).Instance()

**Reply by hwbrzzl** [reply]
You can use `facades.Orm.Query().(*gorm.QueryImpl).Instance`, but we don't recommend it.

**Comment by haodiao** [comment]
```go
facades.Orm().Query().Model(&models.User{}).Where("id", 123).Update(models.User{Status: 0})
```
数据库 status 这个字段，使用这个方法无法更新的吗

**Reply by hwbrzzl** [reply]
Here: https://www.goravel.dev/zh/orm/getting-started.html#%E6%9B%B4%E6%96%B0%E5%8D%95%E4%B8%80%E5%AD%97%E6%AE%B5

<img width="792" alt="image" src="https://github.com/user-attachments/assets/1ae06e8d-d482-4125-8114-170487dff5c1">

**Reply by haodiao** [reply]
##### 好吧，是我没怎么注意，我之前更新了数据库发现没生效。    
也借此给一些建议 (就不去 github 上提 issues 了)：
1、建议默认添加上进程的终止信号，否则运行定时任务时，重启一次等太久了，我是在 main 里前后加上如下，应该没有错吧 ：
```go
	killSignal := make(chan os.Signal, 1)
	signal.Notify(killSignal, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGUSR1, syscall.SIGUSR2)
	//......
	select {
	case <-killSignal:
		log.Printf("stop signal caught, pid[%d] stopping...", os.Getpid())
	}
```
2、建议加上个 websocket，我是用 gorilla/websocket 在某个url路径上实现 (不单独开端口)  
3、laravel 新版本已经默认不用  kernel 了，比如定时任务的 kernel 改为 routes 的 console 文件（虽然影响不大）。  
4、建议把所要能升级的包都升级到最新版本（我是喜欢用 go get -u all && go mod tidy来更新）

**Reply by hwbrzzl** [reply]
感谢建议：
1. 有一个 issue 在跟踪：https://github.com/goravel/goravel/issues/435
2. 暂时没有计划加入框架，但有一个集成的例子：https://github.com/goravel/example?tab=readme-ov-file#websocket
3. 之后会优化一下
4. 框架大版本升级时候会升级所有包

**Comment by XarchencoMO** [comment]
Can i doing something like this: 

user := User{}.First()
users := User{}.Where(// condition).Get()
users := User{}.All()

and other similar designs ?

**Reply by hwbrzzl** [reply]
Not yet.

**Comment by ruoxiao-zh** [comment]
Does it support TiDB database

**Reply by hwbrzzl** [reply]
Based on the Gorm document, should support, you can try it: https://gorm.io/docs/connecting_to_the_database.html#TiDB

**Reply by devhaozi** [reply]
> Does it support TiDB database

Supported, use mysql driver.

**Comment by klphp** [comment]
Events和观察者有demo吗，不知道为什么我设置完creating不触发。是不是设置完事件或观察者后执行
```php 
facades.Orm().Query().Create(&user)
```
应该自动触发的？

**Reply by hwbrzzl** [reply]
是的，会触发 creating，可以贴出你的完整代码看下，或者提交复现 PR 到 goravel/example。

**Comment by jxgame** [comment]
```
facades.Orm().Query().Select("name", "age").Get(&users)
```
这个方法，输出的结果仍然是整个表结构体，只是将非Select里面的字段默认值了，这是合理的吗？

**Reply by krishankumar01** [reply]
The fields not included in the `Select` will still have their zero values in the struct, as that’s how Golang handles struct fields by default.

**Comment by tekpriest** [comment]
I have this bug in `UpdateOrCreate`

```go
    var invitation models.Invitation
	if err := facades.Orm().Query().UpdateOrCreate(
		&invitation,
		models.Invitation{
			Email:  strings.ToLower(data.Email),
			TeamID: teamID,
		},
		models.Invitation{
		Expiry:  carbon.Now().AddDays(3).ToDateTimeStruct(),
		Role:    models.MemberRole(data.Role),
		Country: data.Country,
		Status:  models.InvitationStatus_PENDING,
			TeamID: teamID,
			Email:  strings.ToLower(data.Email),
		},
	); err != nil {
		return err
	}
```
It works if the entry doesn't exist in the database but if it does, it updates which is expected but then also creates after which is not expected

```bash
[...]/go/pkg/mod/github.com/goravel/framework@v1.14.7/database/gorm/query.go:762
[0.318ms] [rows:1] SELECT * FROM "invitations" WHERE ("invitations"."email" = 'asajuenitan+1@gmail.com' AND "invitations"."teamId" = '949789e4-4731-482c-98f1-fd05c2ef3fa4') AND "invitations"."deletedAt" IS NULL ORDER BY "invitations"."id" LIMIT 1

[...]/go/pkg/mod/github.com/goravel/framework@v1.14.7/database/gorm/query.go:621
[0.608ms] [rows:0] UPDATE "invitations" SET "expiry"='2024-11-20 02:23:49.505',"updatedAt"='2024-11-17 02:23:49.505',"createdAt"='2024-11-17 04:22:24.134',"deletedAt"=NULL,"email"='asajuenitan+1@gmail.com',"role"='OWNER',"country"='NG',"status"='PENDING',"teamId"='949789e4-4731-482c-98f1-fd05c2ef3fa4' WHERE "invitations"."deletedAt" IS NULL AND "id" = '1f7e1705-bf70-4de9-93c4-ec3b05479bf2'

[...]/tp/go/pkg/mod/github.com/goravel/framework@v1.14.7/database/gorm/query.go:1317
[1.017ms] [rows:1] INSERT INTO "invitations" ("expiry","updatedAt","createdAt","deletedAt","email","role","country","status","id","teamId") VALUES ('2024-11-20 02:23:49.505','2024-11-17 03:23:49.505','2024-11-17 04:22:24.134',NULL,'asajuenitan+1@gmail.com','OWNER','NG','PENDING','1f7e1705-bf70-4de9-93c4-ec3b05479bf2','949789e4-4731-482c-98f1-fd05c2ef3fa4') ON CONFLICT ("id") DO UPDATE SET "updatedAt"='2024-11-17 02:23:49.506',"expiry"="excluded"."expiry","deletedAt"="excluded"."deletedAt","email"="excluded"."email","role"="excluded"."role","country"="excluded"."country","status"="excluded"."status","teamId"="excluded"."teamId"
```

**Reply by hwbrzzl** [reply]
I didn't find `"id" = '1f7e1705-bf70-4de9-93c4-ec3b05479bf2'` in your code, but it exists in your SQL, missing code?

**Reply by tekpriest** [reply]
oh, nice catch, let me confirm this again

**Reply by tekpriest** [reply]
this was an issue with my implementation.

**Comment by jxgame** [comment]
数据是否存在。
var exists bool
facades.Orm().Query().Where("name", "tom").Exists(&exists)

文档中这里是不是缺少了.Model()

**Reply by jxgame** [reply]
var count int
facades.Orm().Query().Where("name = ?", "tom").Count(&count)
检索聚合 这里似乎也少了.Model()

**Reply by jxgame** [reply]
这个 var count int 应该是。var count int64

**Reply by krishankumar01** [reply]
Yes, these methods should be used with `Table` and `Model` method.

**Reply by hwbrzzl** [reply]
Good catch, updated.

**Comment by HeldvonKosmos** [comment]
Thank you for the Framework!
I like to use Raw SQL Queries, but find only one method in the docs.
Maybe you can complete it like so?

https://laravel-docs.readthedocs.io/en/latest/database/

The insert, update, etc. statement checks if it is what it needs to be for preventing SQL-Injection.

**Reply by hwbrzzl** [reply]
Yes, there is a related issue: https://github.com/goravel/goravel/issues/358

**Reply by hwbrzzl** [reply]
Welcome PR

**Comment by neilyoung** [comment]
I'm unable to create just a simple table:

```
package models

import (
	"github.com/goravel/framework/database/orm"
)

type User struct {
	orm.Model
	Name   string
	Avatar string
	orm.SoftDeletes
}

func (r *User) Connection() string {
  return "mysql"
}

```
```
go run . artisan make:migration create_users_table
go run . artisan migrate
```

Result is missing some columns:

```
mysql> DESCRIBE users;
+------------+-----------------+------+-----+---------+----------------+
| Field      | Type            | Null | Key | Default | Extra          |
+------------+-----------------+------+-----+---------+----------------+
| id         | bigint unsigned | NO   | PRI | NULL    | auto_increment |
| created_at | datetime(3)     | NO   | MUL | NULL    |                |
| updated_at | datetime(3)     | NO   | MUL | NULL    |                |
+------------+-----------------+------+-----+---------+----------------+
```

**Reply by hwbrzzl** [reply]
Do you modify the migration file? `database/migration/***.sql`

**Reply by neilyoung** [reply]
No

**Reply by hwbrzzl** [reply]
It's the reason, you should modify it according to your model.

**Reply by neilyoung** [reply]
Really? Where is this documented? This is different to Laravel

**Reply by neilyoung** [reply]
Oh, I see. It's by design. OK, unexepcted.

**Reply by neilyoung** [reply]
The weird thing is: Why do I need 1000 migrations to add two columns? The more that `add_avatar_to_users_table` doesn't add the avatar column, but the comment column (sigh!)

https://www.goravel.dev/zh/orm/migrations.html#%E7%94%9F%E6%88%90%E8%BF%81%E7%A7%BB

Pretty high class bullshit IMHO. Will have to stick with PHP as backend then, because --- really?

**Reply by neilyoung** [reply]
But I'm not sure if this is really the way to go. Plain GORM does it w/o having to mess with SQL statements.

**Reply by hwbrzzl** [reply]
Sorry, don't get you. But the SQL migration way will be deprecated in v1.16, and the Golang migration feature has been added in v1.15, v1.15 will be released soon.

**Reply by hwbrzzl** [reply]
The new migration style will be:

<img width="631" alt="image" src="https://github.com/user-attachments/assets/22ee6b4a-39f1-4334-92d1-052504f7e8cb" />

**Reply by neilyoung** [reply]
> Sorry, don't get you. But the SQL migration way will be deprecated in v1.16, and the Golang migration feature has been added in v1.15, v1.15 will be released soon.

I mean, a Laravel migration creates a PHP file and you just modify the model, not the SQL. However, your new style looks close to what Laravel does. Will wait for the new version then (the more that I already have married it with Inertia and Vue...)

**Comment by HeldvonKosmos** [comment]


**Reply by neilyoung** [reply]
I think that shouldn't be necessary, but thanks for sharing "Held des Kosmos" :)

**Reply by HeldvonKosmos** [reply]
I really like the approach from atlas to create migrations (also written in go).
https://atlasgo.io/versioned/intro

**Reply by HeldvonKosmos** [reply]
It is much cleaner to use only atlasgo.

**Reply by neilyoung** [reply]
Well, yes, yet another dependency with low life time expectation...

**Comment by neilyoung** [comment]
@hwbrzzl When do you think to publish this version?

**Reply by hwbrzzl** [reply]
Next week

**Reply by neilyoung** [reply]
Perfect

**Comment by neilyoung** [comment]
Thanks for providing the update and Happy New Year. I still have my problems with the database handling. Is it still not required to just change the model, create a migration and run the migration in order to achieve a creation or change of a table?

I was using the model and added a field and created a migration. The script created had nothing to do with what my new user table is supposed to contain. Also the migration was not executed at all..

```
go run . artisan migrate
  INFO    Nothing to migrate
 SUCCESS  Migration success
```

What am I doing wrong?

PS: Oh I see, now I'm having to register the migrations... Sorry, that is weird.

**Comment by wangxuancheng-dev** [comment]
能加一个laravel的查询全局作用域就好了

**Reply by hwbrzzl** [reply]
Had an issue: https://github.com/goravel/goravel/issues/90

**Comment by MeNavel** [comment]
For models with a date data_type, is it written as date or is there another format? because in documentation there is only int and string

**Reply by hwbrzzl** [reply]
You can use any type you want.

**Reply by MeNavel** [reply]
is there any documentation for this? because i just used go after using laravel

**Comment by jxgame** [comment]
一个奇怪的问题，突发，更新数据的时候，有一个表自动将created_at设置为NULL 所有地方都没有对created_at进行过操作。表结构也正常和其它表一样，其它表使用相同的方法也能正常的。代码如下：
···
	//添加或更新familynew
	params := models.FamilyNew{
		Type:        uint(idx),
		TreeID:      uint(tree_id),
		UserId:      uint(user_id),
		MemberID:    uint(member_id),
		Title:       title,
		Description: ctx.Request().Input("description"),
		Content:     content,
	}

	if id > 0 {
		params.ID = uint(id)
	}

	if id == 0 && member.ID == uint(member_id) {
		params.UserId = uint(user_id)
	}

	if params.ID > 0 {
		facades.Orm().Query().Save(&params)
	} else {
		facades.Orm().Query().Create(&params)
	}
···

创建数据时，created_at正常。更新数据它就会将它弄成NULL这是为什么呀，其它表正常只有这表出这个问题。

**Reply by hwbrzzl** [reply]
信息太少了，无法判断，可以本地 debug 一下，或者提交复现 PR 到 https://github.com/goravel/example，我看下。

**Reply by jxgame** [reply]
就是我要更新一个表。
params := models.FamilyNew{Content:"123"}  
facades.Orm().Query().Save(&params) //就这样两句话。
然后，控制台SQL里面却有
update table Set created_at=NULL,content=123 
这样，突然给created_at 设置成了NULL 
从哪里给加的呀？只有一张表这样子，其它表也不这样子。

**Comment by omiyz** [comment]
子查询可以使用吗?
Table()函数中按照gorm的使用方式不行: facades.Orm().Query().Table(`(?) as tb`, query).xxx

**Reply by hwbrzzl** [reply]
信息太少，可以提交一个 issue 详细描述看下。

**Comment by ARSENAL91** [comment]
query := facades.Orm().Query().Model(&models_wages.SumWages{})

	// 条件筛选

	// 使用WHERE IN查询tmp_pg_relation表中的字段
	if employeesId > 0 || beginMonth != "" || endMonth != "" {
		subQuery := facades.Orm().Query().Table("tmp_pg_relation").Select("id")
		if beginMonth != "" {
			subQuery = subQuery.Where("the_month >= ?", beginMonth)
		}
		if endMonth != "" {
			subQuery = subQuery.Where("the_month <= ?", endMonth)
		}
		if employeesId > 0 {
			subQuery = subQuery.Where("employees_id = ?", employeesId)
		}
		query = query.Where("pg_id in (?)", subQuery)
	}
这种子查询，目前看并不支持，请问有什么好的方法吗？

**Reply by hwbrzzl** [reply]
需要拆成两个 sql 去请求。

**Reply by ARSENAL91** [reply]
如果子查询里面的数据太大，放出来容易引起sql过长。看来只能拼接sql了

**Reply by almas-x** [reply]
> 如果子查询里面的数据太大，放出来容易引起sql过长。看来只能拼接sql了

试试  `query.Where("pg_id in (?)", gorm.Expr(subQuery.ToRawSql().Find(map[string]interface{}{})))`

**Reply by ARSENAL91** [reply]
@almas-x,这种也是分开两次sql查询，而且条件也没带进去

**Reply by almas-x** [reply]
> @almas-x,这种也是分开两次sql查询，而且条件也没带进去

这个是拼接带子查询的 SQL 语句的，并不是查两次，注意一下 subQuery 在调用 Find 前是有调 ToRawSql

**Comment by ARSENAL91** [comment]
你好，我这边的业务需要分库，然后是在同一个实例中。我发现，只要我的model定义在非default的那个2库中，每次查询2库，都会产生一个新的连接，这样就很容易导致"Too many connections"问题，而循环查询default库则不会。请问是什么问题吗。配置代码如下

"default": config.Env("DB_CONNECTION", "mysql"),
		// Database connections
		"connections": map[string]any{
			"mysql": map[string]any{
				"driver":   "mysql",
				"host":     config.Env("DB_HOST", "127.0.0.1"),
				"port":     config.Env("DB_PORT", 3306),
				"database": config.Env("DB_DATABASE", "forge"),
				"username": config.Env("DB_USERNAME", ""),
				"password": config.Env("DB_PASSWORD", ""),
				"charset":  "utf8mb4",
				"loc":      "Local",
				"prefix":   "",
				"singular": false,
			},
			"ybs_mysql": map[string]any{
				"driver":   "mysql",
				"host":     config.Env("DB_HOST", "127.0.0.1"),
				"port":     config.Env("DB_PORT", 3306),
				"database": config.Env("DB_YBS_DATABASE", "forge"),
				"username": config.Env("DB_USERNAME", ""),
				"password": config.Env("DB_PASSWORD", ""),
				"charset":  "utf8mb4",
				"loc":      "Local",
				"prefix":   "",
				"singular": false,
			},
			},
		},

**Reply by ARSENAL91** [reply]
是同一次请求，多次请求ybs_mysql库，则都会产生多次连接

**Reply by ARSENAL91** [reply]
已解决，是需要在定义model前加上连接的库
facades.Orm().Connection("ybs_mysql").Query().

**Comment by Benjababe** [comment]
Hi, I was wondering if it's possible to change the behaviour of the SQL commands being printed onto the console? Like hiding successful SQL queries and only printing queries returning an error. 

Or at least set a `LogLevel` like [GORM does](https://gorm.io/docs/logger.html)

**Reply by hwbrzzl** [reply]
You can set the configuration `app.debug` to `false`, then only warning, error will be printed.

Or you can set the configuration `logging.channels.single/daily.print` to `false`, then the console will not print logs anymore, only print logs in the storage log file.
